specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,257 @@
+
+# Specification
+
+## Message Schemas
+
+Message body schemas are notated like this:
+
+```
+(
+    field_name_1: type
+    field_name_2: type
+)
+```
+
+Embedded types are serialized as SSZ Containers unless otherwise noted.
+
+All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.
+
+## `libp2p` Protocol Names
+
+A "Protocol ID" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. Peers exchange supported protocol IDs via the `Identify` protocol upon connection. When opening a new stream, peers pin a particular protocol ID to it, and the stream remains contextualised thereafter. Since messages are sent inside a stream, they do not need to bear the protocol ID.
+
+## RPC-Over-`libp2p`
+
+To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/beacon/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:
+
+```
+(
+    id: uint64
+    method_id: uint16
+    body: Request
+)
+```
+
+and their corresponding responses are wrapped in a "response" structure:
+
+```
+(
+    id: uint64
+    is_error: boolean
+    result: Response
+)
+```
+
+If an error occurs, a variant of the response structure is returned:
+
+```
+(
+    id: uint64
+    is_error: boolean
+    result: (
+        code: uint16
+        data: bytes
+    )
+)
+```
+
+The details of the RPC-Over-`libp2p` protocol are similar to [JSON-RPC 2.0](https://www.jsonrpc.org/specification). Specifically:
+
+1. The `id` member is REQUIRED.
+2. The `id` member in the response MUST be the same as the value of the `id` in the request.
+3. The `id` member MUST be unique within the context of a single connection. Monotonically increasing `id`s are RECOMMENDED.
+4. The `method_id` member is REQUIRED.
+5. The `result` member is required on success, and MUST NOT exist if there was an error.
+6. The `error` member is REQUIRED on errors, and MUST NOT exist if there wasn't an error.
+7. `is_error` MUST be `true` on errors, or `false` otherwise.
+
+Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching. Note that this implies that responses MAY arrive in a different order than requests.
+
+The "method ID" fields in the below messages refer to the `method` field in the request structure above.
+
+The first 1,000 values in `error.code` are reserved for system use. The following error codes are predefined:
+
+1. `0`: Parse error.
+2. `10`: Invalid request.
+3. `20`: Method not found.
+4. `30`: Server error.
+
+## Messages
+
+### Hello
+
+**Method ID:** `0`
+
+**Body**:
+
+```
+(
+    network_id: uint8
+    latest_finalized_root: bytes32
+    latest_finalized_epoch: uint64
+    best_root: bytes32
+    best_slot: uint64
```

Comment:
slots are based on wall time - what's the best_slot field for?

Reply:
pretty sure this is supposed to refer to the slot of the head block. Maybe rename `best_root` and `best_slot` to `head_root` and `head_slot` (or to be even more clear `head_block_root/slot`)?

Reply:
I think `head_block_root` and `head_slot` would be clearer